### DISCUSSION TITLE
Duda test callback

### DISCUSSION THREAD

**Discussion by SantiagoCarnezis** [dds-jv/foro#242] [Consultas de objetos] [answered]
Buenas, tengo una duda sobre como testear las funciones asyncRead y asyncWrite del TPI. A mi se me ocurrió mockear el callback y cuando invoque la función asyncRead fijarme si se acepta el callback, ósea fijarme si se hace callback.accept(...).
Para eso yo pensé en mockear el lowFile para que al invocar asyncRead lo que haga es aceptar el callback que le paso por parámetro y luego el final del test verificar que se haya hecho un callback.accept(). 
Pero estoy teniendo un problema, ya que al hacer doAnswer(...).when(...) me tira un error de compilación, dice que requiere un tipo T, pero obtiene un tipo void. Cuando justamente lo que yo quiero hacer dentro del when es invocar un método que de tipo void. 
Para hacer este test yo me guie con lo que dice en la [guia rapida de mockito](https://docs.google.com/document/d/1467Gc-adARJZZhVAdgazdCeHWRzCUJg6CfMD3nkhmG4/edit) en la parte de Stubbing de efectos sin retorno.

**Comment by nanderson87** [comment, marked as answer]
Hola @SantiagoCarnezis 

Te animas a pasarnos una captura o el extracto del test donde estabas tratando de hacer esto y el error de compilación que te tira?

**Reply by SantiagoCarnezis** [reply]
![image](https://user-images.githubusercontent.com/62964148/119927572-09584200-bf50-11eb-9b9f-fe69eef99287.png)
Eso es lo que me tira. Yo lo que quiero hacer es que al pasarle _asyncReadFile_ al _lowLevelFileSystem_, me haga el _aceppt_ del _callback_ que le estoy pasando. Pero se ve que _when_ necesita que se retorne algo. Si cambio el tipo de _asyncReadFile_ a tipo int, me deja de tirar ese error

**Reply by nanderson87** [reply]
Ok, ahí estuve chequeando un poco. Te dejo un ejemplo de como se podría para evaluar el callback. Después lo podes adapatar mejor al test que queres realizar.

```java
final int bytesLeidos = 5;

final Consumer<Integer> consumerMock = mock(Consumer.class);

doAnswer(invocation -> {
  final Consumer<Integer> callback = invocation.getArgument(4);
  callback.accept(bytesLeidos);
  return null;
}).when(lowLevelFileSystem).asyncReadFile(anyInt(), any(), anyInt(), anyInt(), any());


lowLevelFileSystem.asyncReadFile(1, new byte[100], 0, 4, consumerMock);

verify(consumerMock, times(1)).accept(bytesLeidos);
```

Al parecer hay un error en el ejemplo de la documentación ya que la sintaxis correcta es: 
`when(lowLevelFileSystem).asyncReadFile(anyInt(), any(), anyInt(), anyInt(), any());`

Y no:
`when(lowLevelFileSystem..asyncReadFile(anyInt(), any(), anyInt(), anyInt(), any()));`

Ahora lo cambio en el documento.

**Comment by SantiagoCarnezis** [comment]
Muchas gracias, probe y me funciono. Sos un genio!

**Reply by nanderson87** [reply]
Genial!
